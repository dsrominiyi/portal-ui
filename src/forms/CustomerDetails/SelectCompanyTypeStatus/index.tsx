import React, { FunctionComponent, Fragment } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import Dropdown from '../../../components/Dropdown';
import MultiChoice from '../../../components/MultiChoice';
import SubQuestion from '../../../components/SubQuestion';

import {
  companyTypeStatusOptions,
  otherCompanyTypeStatusOptions
} from '../../../helpers/optionLists';

export interface Props {
  typeStatus?: CompanyTypeStatus;
  hasError?: boolean;
  onChange(value: CompanyTypeStatus): void;
}

const standardTypeStatuses = [...companyTypeStatusOptions];
const typeStatusOther = standardTypeStatuses.pop(); // intentionally remove 'Other' from standardTitles

export const SelectCompanyTypeStatus: FunctionComponent<Props> = ({
  typeStatus = '' as CompanyTypeStatus,
  hasError,
  onChange
}: Props) => (
  <Fragment>
    <QuestionBox
      hasError={hasError}
      errorText="Please confirm legal status"
      text="Company legal status"
      inputFields={[
        <MultiChoice
          key="0"
          options={companyTypeStatusOptions}
          selectedOption={
            !typeStatus || standardTypeStatuses.includes(typeStatus) ? typeStatus : typeStatusOther
          }
          onClick={(option: CompanyTypeStatus) => onChange(option)}
          attributes={{ 'test-id': 'select-standard-legal-status' }}
        />
      ]}
    />
    <SubQuestion
      visible={!!typeStatus && !standardTypeStatuses.includes(typeStatus)}
      items={[
        <QuestionBox
          key="0"
          hasError={hasError}
          inputFields={[
            <Dropdown
              key="0"
              label="Please select legal status"
              options={otherCompanyTypeStatusOptions}
              selected={otherCompanyTypeStatusOptions.includes(typeStatus) ? typeStatus : undefined}
              onChange={value => onChange(value as CompanyTypeStatus)}
              attributes={{ 'test-id': 'select-alternative-legal-status' }}
            />
          ]}
        />
      ]}
    />
  </Fragment>
);

SelectCompanyTypeStatus.displayName = 'SelectCompanyTypeStatus';
export default SelectCompanyTypeStatus;
