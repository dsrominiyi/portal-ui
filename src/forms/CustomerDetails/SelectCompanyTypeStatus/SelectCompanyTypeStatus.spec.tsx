import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import SelectCompanyTypeStatus, { Props } from '.';

let fields: ShallowWrapper<Props>;

const onChange = jest.fn();
let standardOptionsMultiChoice: any;
let otherOptions: any;
let otherOptionsDropdown: any;

const update = () => {
  standardOptionsMultiChoice = (fields.props() as any).children[0].props.inputFields[0];
  otherOptions = (fields.props() as any).children[1];
  otherOptionsDropdown = otherOptions.props.items[0].props.inputFields[0];
};

describe('SelectCompanyTypeStatus', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fields = shallow(
      <SelectCompanyTypeStatus typeStatus={CompanyTypeStatus.Limited} onChange={onChange} />
    );
    update();
  });

  test('SelectCompanyTypeStatus matches snapshot', () => {
    expect(fields).toMatchSnapshot();
  });

  test('SelectCompanyTypeStatus with alt typeStatus matches snapshot', () => {
    fields.setProps({ typeStatus: CompanyTypeStatus.AssociationClubSociety });

    expect(fields).toMatchSnapshot();
  });

  test('Selecting a type status calls onChange with the value', () => {
    const typeStatus = CompanyTypeStatus.Limited;
    standardOptionsMultiChoice.props.onClick(typeStatus);

    expect(onChange).toHaveBeenCalledWith(typeStatus);
  });

  test('selecting the Other type status option renders the alternate options dropdown', () => {
    expect(otherOptions.props.visible).toBe(false);

    fields.setProps({ typeStatus: CompanyTypeStatus.Other });
    update();

    expect(otherOptions.props.visible).toBe(true);
  });

  test('Selecting an alternate type status calls onChange with the value', () => {
    const typeStatus = CompanyTypeStatus.AssociationClubSociety;
    otherOptionsDropdown.props.onChange(typeStatus);

    expect(onChange).toHaveBeenCalledWith(typeStatus);
  });
});
