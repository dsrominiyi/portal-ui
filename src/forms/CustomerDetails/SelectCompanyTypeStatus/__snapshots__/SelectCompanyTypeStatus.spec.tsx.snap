// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectCompanyTypeStatus SelectCompanyTypeStatus matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    errorText="Please confirm legal status"
    inputFields={
      Array [
        <MultiChoice
          attributes={
            Object {
              "test-id": "select-standard-legal-status",
            }
          }
          onClick={[Function]}
          options={
            Array [
              "Sole Proprietor",
              "Partnership",
              "Limited",
              "Public Limited",
              "Other",
            ]
          }
          selectedOption="Limited"
        />,
      ]
    }
    text="Company legal status"
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <Dropdown
                attributes={
                  Object {
                    "test-id": "select-alternative-legal-status",
                  }
                }
                label="Please select legal status"
                onChange={[Function]}
                options={
                  Array [
                    "Association/Club/Society",
                    "Charity",
                    "Government",
                    "Limited Liability Partnership",
                    "Partnership - Limited",
                    "Partnership - Ordinary",
                    "Private Unlimited Company",
                    "Trust or Trustees",
                  ]
                }
              />,
            ]
          }
        />,
      ]
    }
    visible={false}
  />
</Fragment>
`;

exports[`SelectCompanyTypeStatus SelectCompanyTypeStatus with alt typeStatus matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    errorText="Please confirm legal status"
    inputFields={
      Array [
        <MultiChoice
          attributes={
            Object {
              "test-id": "select-standard-legal-status",
            }
          }
          onClick={[Function]}
          options={
            Array [
              "Sole Proprietor",
              "Partnership",
              "Limited",
              "Public Limited",
              "Other",
            ]
          }
          selectedOption="Other"
        />,
      ]
    }
    text="Company legal status"
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <Dropdown
                attributes={
                  Object {
                    "test-id": "select-alternative-legal-status",
                  }
                }
                label="Please select legal status"
                onChange={[Function]}
                options={
                  Array [
                    "Association/Club/Society",
                    "Charity",
                    "Government",
                    "Limited Liability Partnership",
                    "Partnership - Limited",
                    "Partnership - Ordinary",
                    "Private Unlimited Company",
                    "Trust or Trustees",
                  ]
                }
                selected="Association/Club/Society"
              />,
            ]
          }
        />,
      ]
    }
    visible={true}
  />
</Fragment>
`;
